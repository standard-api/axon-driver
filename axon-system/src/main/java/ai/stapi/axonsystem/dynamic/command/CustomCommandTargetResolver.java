package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.AbstractCommand;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.modelling.command.AnnotationCommandTargetResolver;
import org.axonframework.modelling.command.CommandTargetResolver;
import org.axonframework.modelling.command.VersionedAggregateIdentifier;
import org.jetbrains.annotations.NotNull;

public class CustomCommandTargetResolver implements CommandTargetResolver {
  
  private final AnnotationCommandTargetResolver annotationCommandTargetResolver;

  public CustomCommandTargetResolver() {
    this.annotationCommandTargetResolver = AnnotationCommandTargetResolver.builder().build();
  }

  @Override
  public VersionedAggregateIdentifier resolveTarget(@NotNull CommandMessage<?> command) {
    var payload = command.getPayload();
    if (payload instanceof AbstractCommand<?> abstractCommand) {
      return new VersionedAggregateIdentifier(
          abstractCommand.getTargetIdentifier(),
          null
      );
    }
    return this.annotationCommandTargetResolver.resolveTarget(command);
  }
}
