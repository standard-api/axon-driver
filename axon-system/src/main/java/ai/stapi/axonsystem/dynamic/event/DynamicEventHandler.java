package ai.stapi.axonsystem.dynamic.event;

import ai.stapi.axonsystem.dynamic.messagehandler.DynamicMessageHandler;
import ai.stapi.graphsystem.messaging.event.DynamicGraphUpdatedEvent;
import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import org.axonframework.eventhandling.EventMessage;

@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.ANNOTATION_TYPE, ElementType.CONSTRUCTOR, ElementType.METHOD})
@DynamicMessageHandler(
    messageType = EventMessage.class,
    payloadType = DynamicGraphUpdatedEvent.class
)
public @interface DynamicEventHandler {

  /**
   * Specifies the name of message that can be handled by the member method.
   */
  String messageName();

}
