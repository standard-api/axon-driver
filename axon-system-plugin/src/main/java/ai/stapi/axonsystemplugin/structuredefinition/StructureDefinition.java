package ai.stapi.axonsystemplugin.structuredefinition;

import ai.stapi.axonsystem.graphaggregate.AggregateWithDynamicGraph;
import ai.stapi.graphoperations.objectGraphMapper.model.MissingFieldResolvingStrategy;
import ai.stapi.graphsystem.dynamiccommandprocessor.DynamicCommandProcessor;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.FixedImportStructureDefinition;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.ImportStructureDefinition;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.StructureDefinitionImported;
import ai.stapi.schema.structuredefinition.StructureDefinitionId;
import ai.stapi.schema.structuredefinition.StructureDefinitionNormalizer;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.eventsourcing.EventSourcingHandler;
import org.axonframework.modelling.command.AggregateCreationPolicy;
import org.axonframework.modelling.command.CreationPolicy;
import org.axonframework.spring.stereotype.Aggregate;
import org.springframework.beans.factory.annotation.Autowired;

@Aggregate
public class StructureDefinition extends AggregateWithDynamicGraph<StructureDefinitionId> {

  @CommandHandler
  @CreationPolicy(AggregateCreationPolicy.CREATE_IF_MISSING)
  public void handle(
      ImportStructureDefinition command,
      @Autowired DynamicCommandProcessor commandProcessor
  ) {
    this.commandProcessor = commandProcessor;
    this.processCommandDynamically(
        new FixedImportStructureDefinition(
            command.getTargetIdentifier(),
            StructureDefinitionNormalizer.normalize(command.getStructureDefinitionSource())
        ),
        MissingFieldResolvingStrategy.LENIENT
    );
  }

  @EventSourcingHandler
  public void on(StructureDefinitionImported event) {
    this.onAggregateCreated(event);
  }
}