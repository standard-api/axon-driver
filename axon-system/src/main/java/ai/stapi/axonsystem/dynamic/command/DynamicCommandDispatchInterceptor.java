package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import java.util.List;
import java.util.function.BiFunction;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.commandhandling.GenericCommandMessage;
import org.axonframework.messaging.GenericMessage;
import org.axonframework.messaging.MessageDispatchInterceptor;
import org.jetbrains.annotations.NotNull;

public class DynamicCommandDispatchInterceptor  implements MessageDispatchInterceptor<CommandMessage<?>> {

  @NotNull
  @Override
  public BiFunction<Integer, CommandMessage<?>, CommandMessage<?>> handle(
      @NotNull List<? extends CommandMessage<?>> messages
  ) {
    return (integer, commandMessage) -> {
      if (commandMessage.getPayload() instanceof DynamicCommand dynamicCommand) {
        return new GenericCommandMessage<>(
            new GenericMessage<>(dynamicCommand),
            dynamicCommand.getSerializationType()
        );
      }
      return commandMessage;
    };
  }
}
