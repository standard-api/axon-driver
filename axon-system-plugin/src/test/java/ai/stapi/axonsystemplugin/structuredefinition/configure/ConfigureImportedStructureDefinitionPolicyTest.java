package ai.stapi.axonsystemplugin.structuredefinition.configure;

import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import ai.stapi.graphsystem.messaging.event.DynamicGraphUpdatedEvent;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.ImportStructureDefinition;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.StructureDefinitionImported;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.identity.UniversallyUniqueIdentifier;
import ai.stapi.schema.structureSchemaProvider.StructureSchemaFinder;
import ai.stapi.schema.structuredefinition.ElementDefinition;
import ai.stapi.schema.structuredefinition.ElementDefinitionType;
import ai.stapi.schema.structuredefinition.RawStructureDefinitionData;
import ai.stapi.schema.structuredefinition.RawStructureDefinitionElementDefinition;
import ai.stapi.schema.structuredefinition.StructureDefinitionId;
import ai.stapi.test.domain.DomainTestCase;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.ActiveProfiles;

@ActiveProfiles("dev")
class ConfigureImportedStructureDefinitionPolicyTest extends DomainTestCase {

  @Autowired
  private StructureSchemaFinder structureSchemaFinder;
  
  @Test
  void itWillConfigureNewStructureDefinition() {
    var testComplex = new RawStructureDefinitionData(
        "TestComplex",
        "http://test.url/for/complex",
        "draft",
        "Structure Definition for test complex ",
        "complex-type",
        false,
        "TestComplex",
        null,
        new RawStructureDefinitionData.Differential(
            new ArrayList<>(
                List.of(
                    new RawStructureDefinitionElementDefinition(
                        "TestComplex.somePrimitiveField",
                        1,
                        "1",
                        "Primitive Example Field on Test Complex",
                        "",
                        "",
                        new ArrayList<>(
                            List.of(
                                new RawStructureDefinitionElementDefinition.ElementDefinitionType(
                                    "string",
                                    new ArrayList<>()
                                )
                            )
                        )
                    )
                )
            )
        )
    );
    this.whenCommandIsDispatched(
        new ImportStructureDefinition(
            new StructureDefinitionId("TestComplex"),
            testComplex
        )
    );
    this.thenExpectedEventTypesSaved(
        StructureDefinitionImported.class,
        StructureDefinitionConfigured.class
    );
    var field = this.structureSchemaFinder.getFieldDefinitionFor(
        "TestComplex",
        "somePrimitiveField"
    );
    this.thenObjectApproved(field);
  }

  @Test
  void itWillConfigureAddedElement() {
    var testComplex = new RawStructureDefinitionData(
        "TestComplex",
        "http://test.url/for/complex",
        "draft",
        "Structure Definition for test complex ",
        "complex-type",
        false,
        "TestComplex",
        null,
        new RawStructureDefinitionData.Differential(
            new ArrayList<>(
                List.of(
                    new RawStructureDefinitionElementDefinition(
                        "TestComplex.somePrimitiveField",
                        1,
                        "1",
                        "Primitive Example Field on Test Complex",
                        "",
                        "",
                        new ArrayList<>(
                            List.of(
                                new RawStructureDefinitionElementDefinition.ElementDefinitionType(
                                    "string",
                                    new ArrayList<>()
                                )
                            )
                        )
                    )
                )
            )
        )
    );
    this.givenCommandIsDispatched(
        new ImportStructureDefinition(
            new StructureDefinitionId("TestComplex"),
            testComplex
        )
    );
    this.whenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("TestComplex"),
            "AddElementOnStructureDefinitionDifferential",
            Map.of(
                "element", new ArrayList<>(List.of(
                    new ElementDefinition(
                        "TestComplex.newField",
                        new ArrayList<>(List.of(new ElementDefinitionType("string"))),
                        0,
                        "1",
                        "New field",
                        "New field",
                        "New field"
                    )
                ))
            )
        )
    );
    this.thenExpectedEventTypesSaved(
        StructureDefinitionImported.class,
        StructureDefinitionConfigured.class,
        DynamicGraphUpdatedEvent.class,
        ElementsToStructureDefinitionConfigured.class
    );
    var field = this.structureSchemaFinder.getFieldDefinitionFor(
        "TestComplex",
        "newField"
    );
    this.thenObjectApproved(field);
  }

  @Test
  @Disabled
  void itCanImportStructureDefinitionAndExecuteAppropriateDynamicCommand() {
    var testPrimitive = new RawStructureDefinitionData(
        "string",
        "http://test.url/for/primitive",
        "draft",
        "Structure Definition for string",
        "primitive-type",
        false,
        "string",
        null,
        new RawStructureDefinitionData.Differential(new ArrayList<>())
    );
    var testComplex = new RawStructureDefinitionData(
        "TestComplex",
        "http://test.url/for/complex",
        "draft",
        "Structure Definition for test complex ",
        "complex-type",
        false,
        "TestComplex",
        null,
        new RawStructureDefinitionData.Differential(
            new ArrayList<>(
                List.of(
                    new RawStructureDefinitionElementDefinition(
                        "TestComplex.somePrimitiveField",
                        1,
                        "1",
                        "Primitive Example Field on Test Complex",
                        "",
                        "",
                        new ArrayList<>(
                            List.of(
                                new RawStructureDefinitionElementDefinition.ElementDefinitionType(
                                    "string",
                                    new ArrayList<>()
                                )
                            )
                        )
                    )
                )
            )
        )
    );
    var testResource = new RawStructureDefinitionData(
        "TestResource",
        "http://test.url/for/Resource",
        "draft",
        "Structure Definition for test Resource ",
        "resource",
        false,
        "TestResource",
        null,
        new RawStructureDefinitionData.Differential(
            new ArrayList<>(
                List.of(
                    new RawStructureDefinitionElementDefinition(
                        "TestResource.somePrimitiveField",
                        1,
                        "1",
                        "Primitive Example Field on Test Resource",
                        "",
                        "",
                        new ArrayList<>(
                            List.of(
                                new RawStructureDefinitionElementDefinition.ElementDefinitionType(
                                    "string",
                                    new ArrayList<>()
                                )
                            )
                        )
                    ),
                    new RawStructureDefinitionElementDefinition(
                        "TestResource.someComplexField",
                        1,
                        "1",
                        "Complex Example Field on Test Resource",
                        "",
                        "",
                        new ArrayList<>(
                            List.of(
                                new RawStructureDefinitionElementDefinition.ElementDefinitionType(
                                    "TestComplex",
                                    new ArrayList<>()
                                )
                            )
                        )
                    )
                )
            )
        )
    );
    var element = new RawStructureDefinitionData(
        "Element",
        "http://test.url/for/Element",
        "draft",
        "Structure Definition for string",
        "complex-type",
        true,
        "Element",
        null,
        new RawStructureDefinitionData.Differential(new ArrayList<>())
    );
    this.givenCommandIsDispatched(
        new ImportStructureDefinition(
            new StructureDefinitionId("Element"),
            element
        )
    );
    this.givenCommandIsDispatched(
        new ImportStructureDefinition(
            new StructureDefinitionId("string"),
            testPrimitive
        )
    );
    this.givenCommandIsDispatched(
        new ImportStructureDefinition(
            new StructureDefinitionId("TestComplex"),
            testComplex
        )
    );
    this.givenCommandIsDispatched(
        new ImportStructureDefinition(
            new StructureDefinitionId("TestResource"),
            testResource
        )
    );
    this.whenCommandIsDispatched(
        new DynamicCommand(
            UniversallyUniqueIdentifier.randomUUID(),
            "CreateTestResource",
            new HashMap<>(
                Map.of(
                    "somePrimitiveField", "Primitive Value",
                    "someComplexField", new HashMap<>(Map.of(
                        "somePrimitiveField", "Example Primitive Value in Complex Value"
                    )
                    )
                )
            )
        ));
    this.thenLastEventOfTypeGraphApproved(DynamicGraphUpdatedEvent.class);
  }
}