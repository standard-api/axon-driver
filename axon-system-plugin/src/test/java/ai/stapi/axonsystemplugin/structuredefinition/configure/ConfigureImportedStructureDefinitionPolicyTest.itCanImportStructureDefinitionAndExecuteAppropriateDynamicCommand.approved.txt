nodes:
	node_type: TestComplex
	node_hash: A711C
	node_edges:
		TestResource -> someComplexField -> TestComplex (14E99 -> 027FF -> A711C)
	attributes:
		somePrimitiveField -> Example Primitive Value in Complex Value

	node_type: TestResource
	node_hash: 14E99
	node_edges:
		TestResource -> someComplexField -> TestComplex (14E99 -> 027FF -> A711C)
	attributes:
		somePrimitiveField -> Primitive Value

edges:
	edge_type: someComplexField
	edge_hash: 027FF
	TestResource -> someComplexField -> TestComplex (14E99 -> 027FF -> A711C)
	attributes:
